tensorflow_quantum/core/ops/noise/tfq_noisy_sampled_expectation.cc

```diff
@@ -0,0 +1,403 @@
+
+        // Use this trajectory as a source for all expectation calculations.
+        for (int j = 0; j < pauli_sums[i].size(); j++) {
+          if (run_samples[j] >= num_samples[i][j]) {
+            continue;
+          }
+          float exp_v = 0.0;
+          OP_REQUIRES_OK(
+              context, ComputeSampledExpectationQsim(pauli_sums[i][j], sim, ss,
+                                                     sv, scratch, 1, &exp_v));
+          rolling_sums[j] += static_cast<double>(exp_v);
+          run_samples[j]++;
+        }
+        bool break_loop = true;
+        for (int j = 0; j < num_samples[i].size(); j++) {
+          if (run_samples[j] < num_samples[i][j]) {
+            break_loop = false;
+            break;
+          }
+        }
+        if (break_loop) {
+          for (int j = 0; j < num_samples[i].size(); j++) {
+            rolling_sums[j] /= num_samples[i][j];
+            (*output_tensor)(i, j) = static_cast<float>(rolling_sums[j]);
+          }
+          break;
+        }
+      }
+    }
+  }
+
+  void ComputeSmall(const std::vector<int>& num_qubits,
+                    const int max_num_qubits,
+                    const std::vector<NoisyQsimCircuit>& ncircuits,
+                    const std::vector<std::vector<PauliSum>>& pauli_sums,
+                    const std::vector<std::vector<int>>& num_samples,
+                    tensorflow::OpKernelContext* context,
+                    tensorflow::TTypes<float, 1>::Matrix* output_tensor) {
+    using Simulator = qsim::Simulator<const qsim::SequentialFor&>;
+    using StateSpace = Simulator::StateSpace;
+    using QTSimulator =
+        qsim::QuantumTrajectorySimulator<qsim::IO, QsimGate,
+                                         qsim::MultiQubitGateFuser, Simulator>;
+
+    const int output_dim_batch_size = output_tensor->dimension(0);
+    std::vector<tensorflow::mutex> batch_locks(output_dim_batch_size,
+                                               tensorflow::mutex());
+
+    const int num_threads = context->device()
+                                ->tensorflow_cpu_worker_threads()
+                                ->workers->NumThreads();
+
+    // [num_threads, batch_size].
+    std::vector<std::vector<int>> rep_offsets(
+        num_threads, std::vector<int>(output_dim_batch_size, 0));
+
+    BalanceTrajectory(num_samples, num_threads, &rep_offsets);
+
+    output_tensor->setZero();
+
+    Status compute_status = Status::OK();
+    auto c_lock = tensorflow::mutex();
+    auto DoWork = [&](int start, int end) {
+      // Begin simulation.
+      const auto tfq_for = qsim::SequentialFor(1);
+      int largest_nq = 1;
+      Simulator sim = Simulator(tfq_for);
+      StateSpace ss = StateSpace(tfq_for);
+      auto sv = ss.Create(largest_nq);
+      auto scratch = ss.Create(largest_nq);
+
+      for (int i = 0; i < ncircuits.size(); i++) {
+        int nq = num_qubits[i];
+        int rep_offset = rep_offsets[start][i];
```

Comment:
no iterations between `start` and `end`. is it intended?

Reply:
Yes, I use a fixed size parallel for block instead of adaptive (like the other ops), so this fine.